src/v/json/iobuf_writer.h

```diff
@@ -73,31 +76,33 @@ class generic_iobuf_writer
         //   3. Drop the final character
         // For each encoded fragment that is written (except the first one):
         //   4. Restore the stashed character over the prefix-quote
-        for (auto i = beg; i != end; ++i) {
+        for (auto i = beg;;) {
```

Comment:
This PR pretty much LGTM, except that I had/am having a hard time convincing myself that this change here is equivalent, and I'm not sure I follow why the restructuring is needed vs just swapping out the last_frag() implementation (also ignoring the early return, which seems fine).

Reply:
Ahh makes sense. Is it generally a bad practice to allow comparing different iterator types? Like could we templatize operator== to support different iterators and just compare their _current (or maybe the template math doesn't math)?

In any case, I stared at it more and yea it does seem equivalent, so lgtm!

Reply:
we have implicit conversion between non-const to const iterator (standard), but I don't think it's generally allowed to go from forward to backward iterator mostly because I don't know how to map the end states in a coherent way

Reply:
Ya I can buy not being able to convert. I'm wondering if implementing a comparator would work

Reply:
I mean semantically how to compare end vs rend?

Reply:
In general C++ doesn't allow comparison of iterators and corresponding reverse iterators, probably for good reasons (reverse iterators are weirder than you might think since you cannot use one-before-the-front in the same way you can use one-past-the-end), so I don't think we should either?

Reply:
>I mean semantically how to compare end vs rend?

Ah, I would've thought something like, end iters aren't equal if their Forward and containers aren't equal.

In any case, not implementing a comparator sgtm -- it does seem like a can of worms, especially if it's not a common practice